### PR TITLE
Refactor pchar_parse() to use switch()

### PR DESCRIPTION
**Pull request by ipylypiv**
Comparisons `c == '$'` and `c == '&' ` are present two times inside if statement.

### CODE REVIEW

**Comment by ctiller** [comment]
Jenkins: this is ok to test

File: src/core/ext/client_channel/uri_parser.c [hunk 1]
```diff
@@ -96,9 +96,9 @@ static size_t parse_pchar(const char *uri_text, size_t i) {
   if (((c >= 'A') && (c <= 'Z')) || ((c >= 'a') && (c <= 'z')) ||
       ((c >= '0') && (c <= '9')) ||
       (c == '-' || c == '.' || c == '_' || c == '~') || /* unreserved */
-      (c == '!' || c == '$' || c == '&' || c == '\'' || c == '$' || c == '&' ||
-       c == '(' || c == ')' || c == '*' || c == '+' || c == ',' || c == ';' ||
-       c == '=') /* sub-delims */) {
+      (c == '!' || c == '$' || c == '&' || c == '\'' ||
```
Comment (ctiller) [inline src/core/ext/client_channel/uri_parser.c]: This is begging to be converted to a switch..

Comment (ctiller) [inline src/core/ext/client_channel/uri_parser.c]: or some form of LUT

**Comment by ctiller** [comment]
Could you run tools/distrib/clang-format-code.sh against this and push an update? The formatting is off compared to what our tooling expects. (you'll need docker installed)

**Comment by ipylypiv** [comment]
Sorry about formatting. Fixed.

File: src/core/ext/client_channel/uri_parser.c (hunk 1 above)
Comment (ipylypiv) [inline src/core/ext/client_channel/uri_parser.c]: Yes, agree. I will work on it. Should I create new pull request for that?

**Comment by ctiller** [comment]
Feel free to amend this one... No need for a new pull request

**Comment by ipylypiv** [comment]
During refactoring I have found few issues:
- missed `'@'` and `':'` characters. By description in comment they are also valid pchars
`pchar = unreserved / pct-encoded / sub-delims / ":" / "@"`
- loop ` for (j = i + 1; j < 2; j++) ` executes only once (but intended to be executed twice)
- check for end of string ` if (uri_text[i + 1] == 0 || uri_text[i + 2] == 0) ` is redundant because 0 ('\0') is not valid hex and this case is handled in the next if statement

Question:
Should we also reject pct-encoded 0 ('\0') byte - `%00`?
[OWASP Null Byte Injection](https://www.owasp.org/index.php/OWASP_Periodic_Table_of_Vulnerabilities_-_Null_Byte_Injection)

Would be one additional check:
` if (uri_text[i + 1] == '0' && uri_text[i + 2] == '0') return NOT_SET;`

**Comment by ipylypiv** [comment]
I am investigating why test failed, and will update my commit to fix that.

**Comment by ctiller** [comment]
This looks good: can you merge with head (the file moved) - and re-run clang-format?
Then I'll trigger tests again and see if we can't merge.

**Comment by ipylypiv** [comment]
Sure, I will merge it.

What about `%00` handling?

> Question:
> Should we also reject pct-encoded 0 ('\0') byte - `%00`?
> [OWASP Null Byte Injection](https://www.owasp.org/index.php/OWASP_Periodic_Table_of_Vulnerabilities_-_Null_Byte_Injection)
> 
> Would be one additional check:
> ` if (uri_text[i + 1] == '0' && uri_text[i + 2] == '0') return NOT_SET;`

**Comment by ctiller** [comment]
We use slices throughout for these things, and I think we have a use case down the line for nulls (anonymous uds sockets on Linux) -- unless there's a better standard representation of these.

**Comment by ctiller** [comment]
Jenkins: test this please

**Comment by ipylypiv** [comment]
I don't correlate latest build failures with my change. Do I need to do something else here?

**Pull request merged**